water_neuron/src/conversion.rs

```diff
@@ -15,6 +20,87 @@ use icrc_ledger_types::icrc2::transfer_from::TransferFromArgs;
 pub const MINIMUM_DEPOSIT_AMOUNT: ICP = ICP::ONE;
 pub const MINIMUM_WITHDRAWAL_AMOUNT: ICP = ICP::from_unscaled(10);
 
+pub async fn cancel_withdrawal(
+    neuron_id: NeuronId,
+) -> Result<MergeResponse, CancelWithdrawalError> {
+    let caller = ic_cdk::caller();
+    let _guard_principal = GuardPrincipal::new(caller)
+        .map_err(|guard_error| CancelWithdrawalError::GuardError { guard_error })?;
+
+    match get_full_neuron(neuron_id.id).await {
+        Ok(result) => match result {
+            Ok(neuron) => match neuron.time_left(timestamp_nanos()) {
+                Some(time) => {
+                    if time < ONE_DAY_SECONDS * 14 {
+                        return Err(CancelWithdrawalError::TooLate);
+                    }
+                }
+                None => return Err(CancelWithdrawalError::UnknownTimeLeft),
+            },
+            Err(gov_err) => return Err(CancelWithdrawalError::GovernanceError(gov_err)),
+        },
+        Err(error) => return Err(CancelWithdrawalError::GetFullNeuronError { message: error }),
+    }
+
+    let icp_due = match read_state(|s| {
+        s.neuron_id_to_withdrawal_id
+            .get(&neuron_id)
+            .and_then(|withdrawal_id| s.withdrawal_id_to_request.get(&withdrawal_id).cloned())
+    }) {
+        Some(withdrawal_request) => {
+            if withdrawal_request.receiver != caller.into() {
+                return Err(CancelWithdrawalError::BadCaller {
+                    message: format!("Caller is not the owner."),
+                });
+            }
+            withdrawal_request.icp_due
+        }
+        None => return Err(CancelWithdrawalError::RequestNotFound),
+    };
+
+    stop_dissolvement(neuron_id)
+        .await
+        .map_err(|error_msg| CancelWithdrawalError::StopDissolvementError { message: error_msg })?;
+
+    match merge_neuron_into_six_months(neuron_id)
```

Comment:
Initially I thought that `merge_neuron_into_six_months` will fail if `disburse` succeeds due to transaction fees, but it doesn't seem to be the case if I am reading code correctly. It would be great to double check that by testing. If merging is guaranteed to fail for a disbursed neuron, then it is easy to prove that there will be no double spending.

Reply:
The 14 days threshold prevents the concurrency between disburse and dissolve now. Do you think it is still necessary to test this?